### DISCUSSION TITLE
Request for 2 new fields on "Get Invoice" response for easier eCommerce integration

### DISCUSSION THREAD

**Discussion by woutersamaey** [btcpayserver/btcpayserver#2908] [General]
Can we get 2 new fields on the "Get Invoice" API response?
- Field `canBeMarkedAsInvalid` as boolean
- Field `canBeMarkedAsSettled` as boolean

This makes it easier on the eCommerce integration to know when an invoice can be manually changed without re-implementing the logic on BTCPay Server's end. This offers greater flexibility and if BTCPay Server were to get more rules and logic (through plugins perhaps?) following this logic would be nearly impossible.

Or another solution:

- Field `availableStatusesForManualMarking` as array, containing "Invalid" and "Settled" (or any other string compatible with the "Mark invoice status" API call.

This would allow a more flexible list of statuses for the future.

**Comment by woutersamaey** [comment]
@Kukks can I get your 2 cents on this please? We could use this to drive the buttons in the Magento backoffice where a shopkeeper can manually mark an order without needing to go to BTCPay Server. This way a customer service employee could handle orders without needing a login on the company's BTCPay Server.

**Comment by woutersamaey** [comment]
The logic from BTCPay Server that the eCommerce site would need to replacate is this complex thing...

```
public bool CanMarkComplete()
{
    return (Status == InvoiceStatusLegacy.Paid) ||
           (Status == InvoiceStatusLegacy.New) ||
           ((Status == InvoiceStatusLegacy.New || Status == InvoiceStatusLegacy.Expired) && ExceptionStatus == InvoiceExceptionStatus.PaidPartial) ||
           ((Status == InvoiceStatusLegacy.New || Status == InvoiceStatusLegacy.Expired) && ExceptionStatus == InvoiceExceptionStatus.PaidLate) ||
           (Status != InvoiceStatusLegacy.Complete && ExceptionStatus == InvoiceExceptionStatus.Marked) ||
           (Status == InvoiceStatusLegacy.Invalid);
}

public bool CanMarkInvalid()
{
    return (Status == InvoiceStatusLegacy.Paid) ||
           (Status == InvoiceStatusLegacy.New) ||
           ((Status == InvoiceStatusLegacy.New || Status == InvoiceStatusLegacy.Expired) && ExceptionStatus == InvoiceExceptionStatus.PaidPartial) ||
           ((Status == InvoiceStatusLegacy.New || Status == InvoiceStatusLegacy.Expired) && ExceptionStatus == InvoiceExceptionStatus.PaidLate) ||
           (Status != InvoiceStatusLegacy.Invalid && ExceptionStatus == InvoiceExceptionStatus.Marked);
}
```

So yeah, having a field like `availableStatusesForManualMarking` would be a great solution.

Code fragements taken from `BTCPayServer.Services.Invoices.InvoiceState`

**Comment by NicolasDorier** [comment]
on it.

**Reply by NicolasDorier** [reply]
https://github.com/btcpayserver/btcpayserver/pull/2950
